core/node/events/stream_ephemeral.go

```diff
@@ -0,0 +1,182 @@
+	if err != nil {
+		if !IsRiverErrorCode(err, Err_NOT_FOUND) {
+			return err
+		}
+
+		// Stream does not exist in the storage - the entire stream is missing.
+		for i := int64(0); i <= lastMiniblockNum; i++ {
+			missingMbs = append(missingMbs, int64(i))
+		}
+	} else if !ephemeral {
+		// Stream exists in the storage and sealed already.
+		return nil
+	} else {
+		// Stream exists in the storage, but not sealed yet, i.e. ephemeral.
+
+		// Get existing miniblock numbers.
+		existingMbs, err := s.params.Storage.ReadEphemeralMiniblockNums(ctx, stream.streamId)
+		if err != nil {
+			return err
+		}
+
+		existingMbsMap := make(map[int64]struct{}, len(existingMbs))
+		for _, num := range existingMbs {
+			existingMbsMap[int64(num)] = struct{}{}
+		}
+
+		for num := int64(0); num <= int64(lastMiniblockNum)+1; num++ {
+			if _, exists := existingMbsMap[num]; !exists {
+				missingMbs = append(missingMbs, num)
+			}
+		}
+	}
+
+	// Fill missing miniblocks
+	if len(missingMbs) > 0 {
+		remotes, _ := stream.GetRemotesAndIsLocal()
+		currentStickyPeer := stream.GetStickyPeer()
+		for range len(remotes) {
+			stub, err := s.params.NodeRegistry.GetNodeToNodeClientForAddress(currentStickyPeer)
+			if err != nil {
+				logging.FromCtx(ctx).Errorw("Failed to get node to node client", "err", err, "streamId", stream.streamId)
+				currentStickyPeer = stream.AdvanceStickyPeer(currentStickyPeer)
+				continue
+			}
+
+			resp, err := stub.GetMiniblocksByIds(ctx, connect.NewRequest[GetMiniblocksByIdsRequest](
+				&GetMiniblocksByIdsRequest{
+					StreamId:     stream.streamId[:],
+					MiniblockIds: missingMbs,
+				},
+			))
+			if err != nil {
+				logging.FromCtx(ctx).Errorw("Failed to get miniblocks from sticky peer", "err", err, "streamId", stream.streamId)
+				currentStickyPeer = stream.AdvanceStickyPeer(currentStickyPeer)
+				continue
+			}
+
+			// Start processing miniblocks from the stream.
+			// If the processing breaks in the middle, the rest of missing miniblocks will be fetched from the next sticky peer.
+			for resp.Receive() {
+				mbInfo, err := NewMiniblockInfoFromProto(resp.Msg().GetMiniblock(), NewParsedMiniblockInfoOpts())
+				if err != nil {
+					return err
+				}
+
+				mbBytes, err := mbInfo.ToBytes()
+				if err != nil {
+					return err
+				}
+
+				if err = s.params.Storage.WriteEphemeralMiniblock(ctx, stream.streamId, &storage.WriteMiniblockData{
+					Number:   mbInfo.Ref.Num,
+					Hash:     mbInfo.Ref.Hash,
+					Snapshot: mbInfo.IsSnapshot(),
+					Data:     mbBytes,
+				}); err != nil {
+					return err
+				}
+
+				// Delete the processed miniblock from the missingMbs slice
+				i := 0
+				mbNum := resp.Msg().GetNum()
+				for _, v := range missingMbs {
+					if v != mbNum {
+						missingMbs[i] = v
+						i++
+					}
+				}
+				missingMbs = missingMbs[:i]
+
+				// No missing miniblocks left, just return.
+				if len(missingMbs) == 0 {
+					return resp.Close()
```

Comment:
Maybe just call close and always return nil? All miniblocks are written and if resp.Close() returns an error it doesn't matter?